File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> pref(n+1, 1);
        vector<int> post(n+1, 1);
        for(auto i=0; i<n; i++){
            pref[i+1] = nums[i] * pref[i];
            post[n-i-1] = post[n-i] * nums[n-i-1];
        }
        vector<int> res(n);
        for(auto i=0; i<n; i++){
            res[i] = pref[i] * post[i+1];
        }
        return res;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        long prod = 1;
        boolean twoZero = false;
        for(int i=0; i<n; i++){
            if(twoZero == true && nums[i] == 0){
                return res;
            }
            if(nums[i] == 0)
                twoZero = true;
            else
                prod = prod * nums[i];
        }
        if(twoZero == true){
            for(int i=0; i<n; i++){
                if(nums[i] == 0)
                    res[i] = (int)prod;
            }
        }else{
            for(int i=0; i<n; i++){
                res[i] = (int)prod / nums[i];
            }
        }
        return res;
    }
}